File: objc/objcsearch/objcsearch/FileTypes.m
#import "config.h"
#import "FileTypes.h"
#import "FileUtil.h"

@interface FileTypes ()


@property NSString *archive;
@property NSString *binary;
@property NSString *code;
@property NSString *text;
@property NSString *unknown;
@property NSString *xml;


@property NSDictionary<NSString*,NSSet<NSString*>*> *fileTypesDict;

@end

@implementation FileTypes

- (instancetype) init {
    self = [super init];
    if (self) {
        self.archive = [NSString stringWithUTF8String:T_ARCHIVE];
        self.binary = [NSString stringWithUTF8String:T_BINARY];
        self.code = [NSString stringWithUTF8String:T_CODE];
        self.text = [NSString stringWithUTF8String:T_TEXT];
        self.unknown = [NSString stringWithUTF8String:T_UNKNOWN];
        self.xml = [NSString stringWithUTF8String:T_XML];
        self.fileTypesDict = [self fileTypesFromJson];
    }
    return self;
}

- (NSDictionary<NSString*,NSSet<NSString*>*>*) fileTypesFromJson {
    NSMutableString *fileTypesJsonPath = [NSMutableString stringWithUTF8String:SHAREDPATH];
    [fileTypesJsonPath appendString:@"/filetypes.json"];

    NSMutableDictionary *fileTypesDict = [[NSMutableDictionary alloc] init];
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:fileTypesJsonPath]) {
        return nil;
    }

    NSData *data = [NSData dataWithContentsOfFile:fileTypesJsonPath];

    if (NSClassFromString(@"NSJSONSerialization")) {
        NSError *error = nil;
        id jsonObject = [NSJSONSerialization
                         JSONObjectWithData:data
                         options:0
                         error:&error];

        if (error) { /* JSON was malformed, act appropriately here */ }

        if ([jsonObject isKindOfClass:[NSDictionary class]]) {
            NSArray *fileTypes = jsonObject[@"filetypes"];
            for (NSDictionary *typeDict in fileTypes) {
                NSArray *extensions = typeDict[@"extensions"];
                NSString *type = typeDict[@"type"];
                fileTypesDict[type] = [NSSet setWithArray:extensions];
            }
        }
    }
    return [NSDictionary dictionaryWithDictionary:fileTypesDict];
}

+ (FileType) fromName:(NSString*)typeName {
    NSString *lname = [typeName lowercaseString];
    if (lname == [NSString stringWithUTF8String:T_TEXT]) {
        return FileTypeText;
    }
    if (lname == [NSString stringWithUTF8String:T_BINARY]) {
        return FileTypeBinary;
    }
    if (lname == [NSString stringWithUTF8String:T_ARCHIVE]) {
        return FileTypeArchive;
    }
    if (lname == [NSString stringWithUTF8String:T_CODE]) {
        return FileTypeCode;
    }
    if (lname == [NSString stringWithUTF8String:T_XML]) {
        return FileTypeXml;
    }
    return FileTypeUnknown;
}

+ (NSString*) toName:(FileType)fileType {
    if (fileType == FileTypeText) {
        return [NSString stringWithUTF8String:T_TEXT];
    }
    if (fileType == FileTypeBinary) {
        return [NSString stringWithUTF8String:T_BINARY];
    }
    if (fileType == FileTypeArchive) {
        return [NSString stringWithUTF8String:T_ARCHIVE];
    }
    if (fileType == FileTypeCode) {
        return [NSString stringWithUTF8String:T_CODE];
    }
    if (fileType == FileTypeXml) {
        return [NSString stringWithUTF8String:T_XML];
    }
    return [NSString stringWithUTF8String:T_UNKNOWN];
}

- (FileType) getFileType:(NSString*)fileName {
    if ([self isCodeFile:fileName]) {
        return FileTypeCode;
    }
    if ([self isXmlFile:fileName]) {
        return FileTypeXml;
    }
    if ([self isTextFile:fileName]) {
        return FileTypeText;
    }
    if ([self isBinaryFile:fileName]) {
        return FileTypeBinary;
    }
    if ([self isArchiveFile:fileName]) {
        return FileTypeArchive;
    }
    return FileTypeUnknown;
}

- (BOOL) isFileOfType:(NSString*)fileName type:(NSString*)typeName {
    NSString *ext = [FileUtil getExtension:fileName];
    return self.fileTypesDict[typeName] != nil &&
    [self.fileTypesDict[typeName] containsObject:ext];
}

- (BOOL) isArchiveFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.archive];
}

- (BOOL) isBinaryFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.binary];
}

- (BOOL) isCodeFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.code];
}

- (BOOL) isTextFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.text];
}

- (BOOL) isXmlFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.xml];
}

- (BOOL) isSearchableFile:(NSString*)fileName {
    return [self getFileType:fileName] != FileTypeUnknown;
}

- (BOOL) isUnknownFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeUnknown;
}

@end

File: swift/swiftsearch/Sources/swiftsearch/FileTypes.swift
//
//  FileTypes.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/12/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation

public enum FileType {
    case unknown, archive, binary, code, text, xml
}

public class FileTypes {
    fileprivate static let archive = "archive"
    fileprivate static let binary = "binary"
    fileprivate static let code = "code"
    fileprivate static let searchable = "searchable"
    fileprivate static let text = "text"
    fileprivate static let unknown = "unknown"
    fileprivate static let xml = "xml"

    private var config: Config
    private var fileTypesDict = [String: Set<String>]()

    public init() {
        self.config = Config()
        // setFileTypesFromXml()
        setFileTypesFromJson()
    }

    private func setFileTypesFromJson() {
        do {
            let fileUrl = URL(fileURLWithPath: config.fileTypesPath)
            let data = try Data(contentsOf: fileUrl, options: .mappedIfSafe)
            if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                if let filetypes = json["filetypes"] as? [[String: Any]] {
                    for ft in filetypes {
                        let typeName = ft["type"] as! String
                        let extensions = ft["extensions"] as! [String]
                        fileTypesDict[typeName] = Set(extensions)
                    }
                    fileTypesDict[FileTypes.text] = fileTypesDict[FileTypes.text]!.union(fileTypesDict[FileTypes.code]!)
                        .union(fileTypesDict[FileTypes.xml]!)
                    fileTypesDict[FileTypes.searchable] =
                        fileTypesDict[FileTypes.text]!.union(fileTypesDict[FileTypes.binary]!)
                            .union(fileTypesDict[FileTypes.archive]!)
                }
            }
        } catch let error as NSError {
            print("Failed to load: \(error.localizedDescription)")
        }
    }

    public static func fromName(_ typeName: String) -> FileType {
        let lname = typeName.lowercased()
        if lname == text {
            return FileType.text
        }
        if lname == binary {
            return FileType.binary
        }
        if lname == archive {
            return FileType.archive
        }
        if lname == code {
            return FileType.code
        }
        if lname == xml {
            return FileType.xml
        }
        return FileType.unknown
    }

    public static func toName(_ fileType: FileType) -> String {
        if fileType == FileType.text {
            return "text"
        }
        if fileType == FileType.binary {
            return "binary"
        }
        if fileType == FileType.archive {
            return "archive"
        }
        if fileType == FileType.code {
            return "code"
        }
        if fileType == FileType.xml {
            return "xml"
        }
        return "unknown"
    }

    public func getFileType(_ fileName: String) -> FileType {
        if isCodeFile(fileName) {
            return FileType.code
        }
        if isXmlFile(fileName) {
            return FileType.xml
        }
        if isTextFile(fileName) {
            return FileType.text
        }
        if isBinaryFile(fileName) {
            return FileType.binary
        }
        if isArchiveFile(fileName) {
            return FileType.archive
        }
        return FileType.unknown
    }

    private func isFileOfType(_ fileName: String, _ typeName: String) -> Bool {
        fileTypesDict.index(forKey: typeName) != nil &&
            fileTypesDict[typeName]!.contains(FileUtil.getExtension(fileName))
    }

    public func isArchiveFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.archive)
    }

    public func isBinaryFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.binary)
    }

    public func isCodeFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.code)
    }

    public func isSearchableFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.searchable)
    }

    public func isTextFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.text)
    }

    public func isUnknownFile(_ fileName: String) -> Bool {
        (fileTypesDict.index(forKey: FileTypes.unknown) != nil &&
            fileTypesDict[FileTypes.unknown]!.contains(FileUtil.getExtension(fileName)))
            || !isSearchableFile(fileName)
    }

    public func isXmlFile(_ fileName: String) -> Bool {
        isFileOfType(fileName, FileTypes.xml)
    }
}
